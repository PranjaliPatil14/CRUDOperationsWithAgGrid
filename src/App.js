import React, { useState, useEffect, useRef } from "react";
import { AgGridReact } from "ag-grid-react";
import "ag-grid-community";
import DeletePost from "./DeletePost";
import BookMarkPost from "./BookMarkPost";
import "./style.css";

export default function App() {
  const [rowData, setRowData] = useState([]);
  const [rowCount, setRowCount] = useState(null);
  const gridApi = useRef(null);

  useEffect(() => {
    fetch("https://jsonplaceholder.typicode.com/posts", { mode: "cors" })
      .then(response => {
        return response.json();
      })
      .then(jsonResponse => {
        setRowData(jsonResponse);
      });
  }, []);

  const onGridReady = params => {
    gridApi.current = params;
  };

  const onModelUpdated = params => {
    setRowCount(params.api.rowModel.rowsToDisplay.length);
  };

  const onAddPost = () => {
    gridApi.current.api.applyTransaction({
      add: [{ title: "Please Enter Post Title...", body: "Please Enter Post Content...", newRow: true }],
      addIndex: 0
    });
  };

  const columnData = [
    {
      headerName: "Post Title",
      field: "title",
      cellClass: "bold-text",
      minWidth: 350
    },
    {
      headerName: "Post Content",
      field: "body"
    },
    {
      headerName: "BookMark",
      field: "bookMark",
      cellRendererFramework: BookMarkPost,
      editable: false,
      sort: "desc",
      filter: false,
      minWidth: 100,
    },
    {
      headerName: "",
      cellRendererFramework: DeletePost,
      editable: false,
      sortable: false,
      filter: false,
      width: 70
    }
  ];
  const defaultConfigs = {
    sortable: true,
    filter: true,
    resizable: true,
    editable: true
  };

  return (
    <div className="page">
    <div className="grid-info">
      <button className="add-post" type="submit" onClick={onAddPost}>
        Add Post
      </button>
        <div><div className="info" /> : Newely Added Posts</div>
        <p className="row-count">
          Total No of Posts: <strong>{rowCount}</strong>
        </p>
        </div>
      <div
        style={{ height: "90vh", width: "100%" }}
        className="ag-theme-alpine"
      >
        <AgGridReact
          columnDefs={columnData}
          rowData={rowData}
          defaultColDef={defaultConfigs}
          onGridReady={onGridReady}
          singleClickEdit
          onModelUpdated={onModelUpdated}
          rowClassRules={{
            "highlight-row": params => params.data.newRow
          }}
        />
      </div>
    </div>
  );
}
