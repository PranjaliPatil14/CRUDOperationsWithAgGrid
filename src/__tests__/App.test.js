import React from "react";
import {
  render,
  waitFor,
  screen,
  fireEvent,
  getByDisplayValue,
} from "@testing-library/react";
import App from "../App";
import postsData from "./test.json";

describe("App", () => {
  const posts = postsData;

  const mockFetchPromise = Promise.resolve({
    json: () => Promise.resolve(posts),
  });

  beforeAll(() => {
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
  });

  afterAll(() => {
    global.fetch.mockClear();
    delete global.fetch;
  });
  test("should render Add Post button, post count and grid with data", async () => {
    const { asFragment } = render(<App suppressColumnVirtualisation />);

    //Wait for grid to get initialized
    await screen.findByText("5");
    // Snapshot of grid info
    expect(asFragment().querySelector(".grid-info")).toMatchSnapshot();

    // Check if data is loaded
    expect(screen.queryByText("Ag Grid Tests")).toBeInTheDocument();
    expect(screen.queryByText("Post Title")).toBeInTheDocument();
    expect(screen.queryByText("Post Content")).toBeInTheDocument();
  });

  test("should add post in the grid and update the total post count", async () => {
    const { container } = render(<App suppressColumnVirtualisation />);

    await screen.findByText("5");
    expect(container.querySelector(".highlight-row")).not.toBeInTheDocument();
    expect(
      screen.queryByText("Please Enter Post Title...")
    ).not.toBeInTheDocument();

    fireEvent.click(screen.queryByText("Add Post"));

    await screen.findByText("6");

    expect(screen.queryByText("6")).toBeInTheDocument();
    expect(
      screen.queryByText("Please Enter Post Title...")
    ).toBeInTheDocument();
    expect(
      screen.queryByText("Please Enter Post Content...")
    ).toBeInTheDocument();
    expect(
      screen.queryByText(new Date().toLocaleDateString())
    ).toBeInTheDocument();

    expect(container.querySelector(".highlight-row")).toBeInTheDocument();
  });

  test("should edit post title to My First blog text", async () => {
    const { container } = render(<App suppressColumnVirtualisation />);

    await screen.findByText("5");

    fireEvent.click(screen.queryByText("Ag Grid Tests"));

    await screen.findByDisplayValue("Ag Grid Tests");

    fireEvent.change(container.querySelector("input[type='text']"), {
      target: { value: "My First Blog" },
    });

    fireEvent.click(screen.queryByText("Add Post"));

    expect(screen.queryByText("My First Blog")).not.toBeNull();
  });
});
