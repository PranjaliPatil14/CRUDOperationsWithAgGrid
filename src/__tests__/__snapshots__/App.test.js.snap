// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render Add Post button, post count and grid with data 1`] = `
<div
  class="grid-info"
>
  <button
    class="add-post"
    type="submit"
  >
    Add Post
  </button>
  <div>
    <div
      class="info"
    />
     : Newely Added Posts
  </div>
  <p
    class="row-count"
  >
    Total No of Posts: 
    <strong>
      5
    </strong>
  </p>
</div>
`;
